src/rust/src/rustzcash.rs

```diff
@@ -449,34 +455,38 @@ pub extern "C" fn librustzcash_sapling_compute_cmu(
     true
 }
 
-/// Computes \[sk\] \[8\] P for some 32-byte point P, and 32-byte Fs.
+/// Computes KDF^Sapling(KA^Agree(sk, P), ephemeral_key).
 ///
 /// If P or sk are invalid, returns false. Otherwise, the result is written to
 /// the 32-byte `result` buffer.
 #[no_mangle]
-pub extern "C" fn librustzcash_sapling_ka_agree(
+pub extern "C" fn librustzcash_sapling_ka_derive_symmetric_key(
     p: *const [c_uchar; 32],
     sk: *const [c_uchar; 32],
+    ephemeral_key: *const [c_uchar; 32],
     result: *mut [c_uchar; 32],
 ) -> bool {
-    // Deserialize p
-    let p = match de_ct(jubjub::ExtendedPoint::from_bytes(unsafe { &*p })) {
-        Some(p) => p,
+    // Deserialize p (either ephemeral_key or pk_d; we can handle them identically).
```

Comment:
```suggestion
    // Deserialize p (representing either epk or pk_d; we can handle them identically).
```
(We use `epk` and `pk_d` for abstract points, whereas `ephemeral_key` or `ephemeralKey` is always a byte sequence. So saying "either ephemeral_key or pk_d" was mixing types.)

Reply:
Note that it isn't a priori obvious that we can handle them identically. $\mathsf{epk}$ has specification type $\mathsf{KA^{Sapling}.Public}$ which is $\mathbb{J}$, with a consensus rule in [§4.5 Output Descriptions](https://zips.z.cash/protocol/protocol.pdf#outputdesc) that excludes small-order points. $\mathsf{pk_d}$, once https://github.com/zcash/zips/issues/664 has been fixed, will have type $\mathsf{KA^{Sapling}.PublicPrimeSubgroup}$ excluding the zero point. That is, $\mathsf{epk}$ allows points of order $\{2, 4, 8\} r_\mathbb{J}$ that are outside the subgroup.

So the documentation "If P or sk are invalid, returns false." must be clarified to say precisely what checks are done on P, so that it is possible to review whether uses of this function are correct in context.